modules/internal/ChapelLocale.chpl

```diff
@@ -247,6 +247,35 @@ module ChapelLocale {
       return "";
     }
 
+    //
+    // Support for different types of memory:
+    // large, random-access (low latency), and streaming (high bandwidth)
+    // Inspired by madvise()'s support for RANDOM and SEQUENTIAL access
+    //
+    pragma "no doc"
+    proc defaultMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc largeMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc randomMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc streamingMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
```

Comment:
Regarding the selection of multiple characteristics, sorry, I tried to take into account all the information from before I got involved, but I somehow missed that.  I am not sure how a hierarchical specification would work.  If someone asked for low latency first, and high bandwidth second, would we give them DDR because latency was first priority, or would we give them MCDRAM because it was the best compromise between the specified characteristics?  If the former, then the KNL locale model would only ever have to look at the first characteristic in order unless a specific size was specified first, and if the latter, then everything except a specific size could be specified with a bit vector rather than a hierarchy.

I think what we are missing most here is experience with a wide range of locale models.  As architectures evolve, it may become clearer what is needed.  I suspect it would be better to start with what we know we are going to need and then grow as requirements change.  I am open to continuing this discussion; the only question is whether it continues as part of this PR or separately.

I have no objection to renaming the characteristics to be about the hardware rather than about the program.